### PR TITLE
feat(apiversion): update apiVersion from v1alpha3 to v1beta1

### LINKED ISSUE #642: Refine the storage spec in HarborCluster CRD to provide more clear and flexible storage configurations
CRD version -> v1beta1?

```YAML
storage:
  kind: filesystem # or oss, swift, gcs, azure etc.
  spec:
    # different spec for different storage driver
    chartPersistentVolume: # Optional
        # Inline the corev1.PersistentVolumeClaimVolumeSource
        # ClaimName is the name of a PersistentVolumeClaim in the same namespace as the pod using this volume.
        # More info: https://kubernetes.io/docs/concepts/storage/persistent-volumes#persistentvolumeclaims
        claimName: myPVC # Required
        # Will force the ReadOnly setting in VolumeMounts.
        readOnly: false # Optional
        prefix: myPrefix # Optional
      registryPersistentVolume: # Optional
        # ... Skipped the same fields with 'chartPersistentVolume': 'claimName', 'readOnly' and 'prefix'.
        # ...
        # Max threads
        maxthreads: 100 # Optional, default = 100, minimal = 25
```

For the current incluster storage, use kind `minio`

```yaml
storage:
  kind: minio/minio
  spec:
      operatorVersion: 4.0.6
    # Image name for the MinIO. It will override the default one.
      image: my-minio # Optional
      # Image pull policy. It will override the global 'imageSource' settings and the default one.
      imagePullPolicy: # Optional, default = IfNotPresent
      # Image pull secrets. It will override the global 'imageSource' settings if it has been set.
      imagePullSecrets:
        - name: myHarborRegSecretOfMinIO
      # Redirection configurations.
      redirect: # Required
        # Determine if the redirection of minio storage is enabled.
        enable: true # Optional, default=true
        # If 'enable' is set to be true, then configure extra settings here.
        # Expose MinIO service for client accessing.
        # Same configuration with the top level 'expose' section.
        expose: # Optional
          # TLS setting
          tls: # Optional
            # Certificate reference
            certificateRef: <cert-ref> # Optional
          # Expose service with ingress way
          ingress:
            # Host of the exposed service
            host: <minio.goharbor.io> # Required
            # Ingress controller type, support ["gce","ncp","default"]
            controller: default # Optional, default value = "default"
            # Annotations applied to the ingress
            annotations: # Optional
              key: value
      # Reference to the secret containing the MinIO access key and secret key.
      secretRef: minioSecret # Optional
      # Supply number of replicas.
      # For standalone mode, supply 1. For distributed mode, supply 4 or more.
      # Note that the operator does not support upgrading from standalone to distributed mode.
      # Specially, 'replicas'*'volumesPerServer' should be >=4.
      replicas: 4 # Required, minimal=1
      # Number of persistent volumes that will be attached per server
      # Specially, 'replicas'*'volumesPerServer' should be >=4.
      volumesPerServer: 1 # Required, minimal=1
      # VolumeClaimTemplate allows a user to specify how volumes inside a MinIOInstance.
      # More info: https://github.com/kubernetes/kubernetes/blob/6b1d87acf3c8253c123756b9e61dac642678305f/pkg/apis/core/types.go#L398
      volumeClaimTemplate: {} # Optional
      # If provided, use these requests and limit for cpu/memory resource allocation
      # More info: https://kubernetes.io/docs/concepts/configuration/manage-compute-resources-container/
      resources: # Optional
        # Limits describes the maximum amount of compute resources allowed.
        # More info: https://kubernetes.io/docs/concepts/configuration/manage-compute-resources-container/
        limits: {} # Optional
        # Requests describes the minimum amount of compute resources required.
        # If Requests is omitted for a container, it defaults to Limits if that is explicitly specified,
        # otherwise to an implementation-defined value.
        # More info: https://kubernetes.io/docs/concepts/configuration/manage-compute-resources-container/
        requests: {} # Optional
```

### LINKED ISSUE #643: Refine the database spec in HarborCluster CRD to provide more clear and flexible db configurations
CRD version -> v1beta1

```yaml
database:
  kind: postgresql
  spec:
    # PostgreSQL user name to connect as.
    # Defaults to be the same as the operating system name of the user running the application.
    username: postsql # Required
    # Secret containing the password to be used if the server demands password authentication.
    passwordRef: psqlSecret # Optional
    # PostgreSQL hosts.
    # At least 1.
    hosts:
      # Name of host to connect to.
      # If a host name begins with a slash, it specifies Unix-domain communication rather than
      # TCP/IP communication; the value is the name of the directory in which the socket file is stored.
      - host: psql # Required
      # Port number to connect to at the server host,
      # or socket file name extension for Unix-domain connections.
      # Zero, specifies the default port number established when PostgreSQL was built.
        port: 5432 # Optional
    # PostgreSQL has native support for using SSL connections to encrypt client/server communications for increased security.
    # Supports values ["disable","allow","prefer","require","verify-ca","verify-full"].
    sslMode: prefer # Optional, default=prefer
    prefix: prefix # Optional
```

or handle in-cluster db powered by zlando/postgresql

```yaml
database:
  kind: zlando/postgresql
  spec:
    operatorVersion: 1.5.0
    # Image name for the PostgresSQL. It will override the default one.
      image: my-psql # Optional
      # Image pull policy. It will override the global 'imageSource' settings and the default one.
      imagePullPolicy: # Optional, default = IfNotPresent
      # Image pull secrets. It will override the global 'imageSource' settings if it has been set.
      imagePullSecrets:
        - name: myHarborRegSecretOfPsql
      # Specify the storage size for the PostgresSQL.
      storage: 1Gi # Optional, default="1Gi"
      # Replicas of PostgresSQL instances.
      replicas: 3 # Optional, default=3
      # The storage class used for creating storage.
      storageClassName: default # Optional
      # If provided, use these requests and limit for cpu/memory resource allocation
      # More info: https://kubernetes.io/docs/concepts/configuration/manage-compute-resources-container/
      resources: # Optional
        # Limits describes the maximum amount of compute resources allowed.
        # More info: https://kubernetes.io/docs/concepts/configuration/manage-compute-resources-container/
        limits: {} # Optional
        # Requests describes the minimum amount of compute resources required.
        # If Requests is omitted for a container, it defaults to Limits if that is explicitly specified,
        # otherwise to an implementation-defined value.
        # More info: https://kubernetes.io/docs/concepts/configuration/manage-compute-resources-container/
        requests: {} # Optional
```

### LINKED ISSUE #644: Refine the redis spec in HarborCluster CRD to provide more clear and flexible configurations
CRD version: v1beta1?

```yaml
redis:
  kind: redis
  spec:
    # Server host.
    host: myredis.com # Required
    # Server port.
    port: 6347 # Required
    # For setting sentinel masterSet.
    sentinelMasterSet: sentinel # Optional
    # Secret containing the password to use when connecting to the server.
    passwordRef: pwdSecret # Optional
    # Secret containing the client certificate to authenticate with.
    certificateRef: cert # Optional
```

Or for the in-cluster redis powered by RedisFailover,

```yaml
redis:
  kind: redisfailover
  spec:
      operatorVersion: 1.0
    # Image name for the Redis. It will override the default one.
      image: my-redis # Optional
      # Image pull policy. It will override the global 'imageSource' settings and the default one.
      imagePullPolicy: # Optional, default = IfNotPresent
      # Image pull secrets. It will override the global 'imageSource' settings if it has been set.
      imagePullSecrets:
        - name: myHarborRegSecretOfRedis
      # Redis sentinel
      sentinel: # Required
        # Replicas of the sentinel service.
        replicas: 3 # Optional, default=3
      # Redis server.
      server: # Required
        # Replicas of the server.
        replicas: 3 # Optional, default=3
        # Storage class used to apply storage of redis.
        StorageClassName: default # Optional
        # Storage size.
        storage: 1Gi # Optional
        # If provided, use these requests and limit for cpu/memory resource allocation
        # More info: https://kubernetes.io/docs/concepts/configuration/manage-compute-resources-container/
        resources: # Optional
          # Limits describes the maximum amount of compute resources allowed.
          # More info: https://kubernetes.io/docs/concepts/configuration/manage-compute-resources-container/
          limits: {} # Optional
          # Requests describes the minimum amount of compute resources required.
          # If Requests is omitted for a container, it defaults to Limits if that is explicitly specified,
          # otherwise to an implementation-defined value.
          # More info: https://kubernetes.io/docs/concepts/configuration/manage-compute-resources-container/
          requests: {} # Optional
```

### PR DESCRIPTION
**Pull request by cndoit18**
update apiVersion from v1alpha3 to v1beta1

Link Issues: helm/helm#9885

closes #643, closes #644, closes #642

Signed-off-by: cndoit18 <cndoit18@outlook.com>

### CODE REVIEW

File: pkg/cluster/controllers/database/utils.go [hunk 1]
```diff
@@ -57,8 +57,8 @@ func (p *PostgreSQLController) GetDatabaseConn(ctx context.Context, ns, secretNa
 
 // GetStorageClass returns the storage class name.
 func (p *PostgreSQLController) GetStorageClass(harborcluster *goharborv1.HarborCluster) string {
-	if harborcluster.Spec.InClusterDatabase != nil && harborcluster.Spec.InClusterDatabase.PostgresSQLSpec != nil {
-		return harborcluster.Spec.InClusterDatabase.PostgresSQLSpec.StorageClassName
+	if harborcluster.Spec.Database.Kind == "Zlando/PostgreSQL" && harborcluster.Spec.Database.Spec.ZlandoPostgreSQL != nil {
```
Comment (steven-zou) [inline pkg/cluster/controllers/database/utils.go:60]: Define  "Zlando/PostgreSQL" as a const value?

File: apis/goharbor.io/v1alpha3/harborcluster_webhook.go [hunk 2]
```diff
@@ -59,7 +59,7 @@ var _ webhook.Validator = &HarborCluster{}
 func (hc *HarborCluster) ValidateCreate() error {
 	clog.Info("validate creation", "name", hc.Name, "namespace", hc.Namespace)
 
-	return hc.validate(nil)
+	return hc.validate(hc)
```
Comment (steven-zou) [inline apis/goharbor.io/v1alpha3/harborcluster_webhook.go:62]: Why we still need v1alpha3 webhook?
Wher is the v1beta1 webhook?

File: pkg/cluster/controllers/harbor/harbor.go [hunk 3]
```diff
@@ -117,7 +114,36 @@ func (harbor *Controller) getHarborCR(ctx context.Context, harborcluster *goharb
...
+			UpdateStrategyType:     spec.UpdateStrategyType,
+			Version:                spec.Version,
+			Expose:                 spec.Expose,
+			HarborComponentsSpec:   spec.HarborComponentsSpec,
+			ImageSource:            spec.ImageSource,
+			Proxy:                  spec.Proxy,
+		},
+	}
+
+	if harborcluster.Spec.Storage.Spec.FileSystem != nil {
```
Comment (steven-zou) [inline pkg/cluster/controllers/harbor/harbor.go:134]: We need to make sure only one storage is defined in the spec via webhook?

**Review by steven-zou — COMMENTED** [review]
some comments

File: pkg/cluster/controllers/harbor/harbor.go (hunk 3 above)
Comment (cndoit18) [reply pkg/cluster/controllers/harbor/harbor.go:134]: If there is already a Kind distinction, it looks like it should not be needed

File: pkg/cluster/controllers/database/utils.go (hunk 1 above)
Comment (cndoit18) [reply pkg/cluster/controllers/database/utils.go]: done

File: apis/goharbor.io/v1alpha3/harborcluster_webhook.go (hunk 2 above)
Comment (cndoit18) [reply apis/goharbor.io/v1alpha3/harborcluster_webhook.go:62]: added in v1beta1/harborcluster

File: pkg/cluster/controllers/harbor/harbor.go (hunk 3 above)
Comment (steven-zou) [reply pkg/cluster/controllers/harbor/harbor.go:134]: As I mentioned, the storage definitions are marked optional with pointers. But the user can also provide more than 1 storage in the spec. e.g:

```
spec:
  kind: s3
  spec:
    filesystem: {},
    s3: {},
    minio: {}
```

It seems we did not use spec.kind in the spec. If we did not add validating logic in the validating webhook, we need to take care of code here. Providing validating webhook is more simple, right?

@cndoit18

Comment (steven-zou) [reply pkg/cluster/controllers/harbor/harbor.go:134]: or:

```yaml
spec:
  kind: s3
 spec:
   minio:
    size: 100GiB
```

**Comment by yxxhero** [comment]
CRD files cannot be templated. They must be plain YAML documents. https://helm.sh/docs/topics/charts/ @cndoit18

**Review by heww — APPROVED** [review]
LGTM

**Review by bitsf — APPROVED** [review]

**Pull request merged**